src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPEBuildManager.cs

```diff
@@ -0,0 +1,462 @@
+            ImmutableHashSet<string> newSharedDesignTimeInputs;
+            var addedDesignTimeInputs = new List<string>();
+            var removedDesignTimeInputs = new List<string>();
+            bool hasRemovedDesignTimeSharedInputs = false;
+            bool hasAddedDesignTimeSharedInputs = false;
+
+            if (value.HasFileChanges)
+            {
+                var designTimeInputs = previousValue.Inputs.ToBuilder();
+                var designTimeSharedInputs = previousValue.SharedInputs.ToBuilder();
+
+                foreach (string item in value.AddedItems)
+                {
+                    // If a property changes we might be asked to add items that already exist, so we just ignore them as we're happy using our existing CancellationSeries
+                    if (designTimeInputs.Add(item))
+                    {
+                        addedDesignTimeInputs.Add(item);
+                    }
+                }
+
+                foreach (string item in value.AddedSharedItems)
+                {
+                    if (designTimeSharedInputs.Add(item))
+                    {
+                        hasAddedDesignTimeSharedInputs = true;
+                    }
+                }
+
+                foreach (string item in value.RemovedItems)
+                {
+                    if (designTimeInputs.Remove(item))
+                    {
+                        removedDesignTimeInputs.Add(item);
+                    }
+                    if (designTimeSharedInputs.Remove(item))
+                    {
+                        hasRemovedDesignTimeSharedInputs = true;
+                    }
+                }
+
+                newDesignTimeInputs = designTimeInputs.ToImmutable();
+                newSharedDesignTimeInputs = designTimeSharedInputs.ToImmutable();
+            }
+            else
+            {
+                // If there haven't been file changes we can just flow our previous collections to the new version to avoid roundtriping our collections to builders and back
+                // Normally ShouldValueBeApplied can be used for this but our project properties require comparison to previous values so we have to apply every update
+                newDesignTimeInputs = previousValue.Inputs;
+                newSharedDesignTimeInputs = previousValue.SharedInputs;
+            }
+
+            // Apply our new value
+            AppliedValue = new ProjectVersionedValue<DesignTimeInputsItem>(new DesignTimeInputsItem
+            {
+                Inputs = newDesignTimeInputs,
+                SharedInputs = newSharedDesignTimeInputs,
+                RootNamespace = value.RootNamespace,
+                OutputPath = value.OutputPath
+            }, value.DataSourceVersions);
+
+
+            // Fire off any events necessary
+
+            // We ignore blank values for previous value so that the initial load of the project doesn't trigger a compile of all PEs. It also makes testing easier.
```

Comment:
This doesn't sound correct, at least for RootNamespace, it's definitely value to go from empty -> value -> empty.

Reply:
I edit the project file and change `<RootNamespace>Foo</RootNamespace>` -> `<RootNamespace></RootNamespace>`. Or I use Property Pages against a CPS project and change Default namespace to empty.

Reply:
> I edit the project file and change <RootNamespace>Foo</RootNamespace> -> <RootNamespace></RootNamespace>

This is the scenario I mentioned above that is broken. After doing this File -> New Class generates a file that doesn't compile (produces `namespace ` as the namespace), as will running the single file generator for a resource file in a subfolder of the project (produces `namespace .FolderName` as the namespace). The single file generator for a `.settings` file in the `Properties` folder didn't run at all.

> Or I use Property Pages against a CPS project and change Default namespace to empty.

This writes `<RootNamespace />` to the project file so results in the same broken experience as above.

Reply:
This has been resolved by changes due to the suggestion ProjectRuleSource hint in https://github.com/dotnet/project-system/pull/4349#discussion_r238940967